addons/mass_mailing/controllers/main.py

```diff
@@ -105,6 +105,35 @@ def track_mail_open(self, mail_id, token, **post):
 
         return response
 
+    @http.route(['/mailing/<int:mailing_id>/view'], type='http', website=True, auth='public')
+    def view(self, mailing_id, email=None, res_id=None, token=""):
+        mailing = request.env['mailing.mailing'].sudo().browse(mailing_id)
+        if mailing.exists():
+            res_id = int(res_id) if res_id else False
+            if not self._valid_unsubscribe_token(mailing_id, res_id, email, str(token)) and not request.env.user.has_group('mass_mailing.group_mass_mailing_user'):
+                raise exceptions.AccessDenied()
+
+            res = mailing.convert_links()
+            base_url = request.env['ir.config_parameter'].sudo().get_param('web.base.url').rstrip('/')
+            urls_to_replace = [
+               (base_url + '/unsubscribe_from_list', mailing._get_unsubscribe_url(email, res_id)),
+               (base_url + '/view', mailing._get_view_url(email, res_id))
+            ]
+            for url_to_replace, new_url in urls_to_replace:
+                if url_to_replace in res[mailing_id]:
+                    res[mailing_id] = res[mailing_id].replace(url_to_replace, new_url if new_url else '#')
+
+            res[mailing_id] = res[mailing_id].replace(
+                'class="o_snippet_view_in_browser"',
+                'class="o_snippet_view_in_browser" style="display: none;"'
```

Comment:
Sad! Can't we think of something cleaner to do this, e.g. with a CSS class that is set higher up in the layout for mailing "view" mode?